### PR TITLE
Add more specs for `Enumerator.product`

### PR DESCRIPTION
**Pull request by AI-Mozi**
Based on the @andrykonchin comment from my last [PR](https://github.com/ruby/spec/pull/1017) I added more specs.

> Enumerator.product has been added. Enumerator::Product is the implementation. [[Feature #18685](https://bugs.ruby-lang.org/issues/18685)]

### CODE REVIEW

File: core/enumerator/product_spec.rb [hunk 1]
```diff
@@ -2,6 +2,29 @@
 
 ruby_version_is "3.2" do
   describe "Enumerator.product" do
+    it "creates a finite enumerator" do
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:5]: **minor:**
```suggestion
    it "creates a Cartesian product of enumerators" do
```

File: core/enumerator/product_spec.rb [hunk 2]
```diff
@@ -2,6 +2,29 @@
...
+      enum = Enumerator.product(1..2, ["A", "B"])
+      enum.to_a.should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"]]
+    end
+
+    it "accepts a list of enumerators of any length" do
+      enum = Enumerator.product(1.., ["A"], ["B"])
+      enum2 = Enumerator.product(2.., ["A"], ["B"], ["C"])
+
+      enum.take(2).should == [[1, "A", "B"], [2, "A", "B"]]
+      enum2.take(2).should == [[2, "A", "B", "C"], [3, "A", "B", "C"]]
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:15]: **minor:** I would avoid infinite enumerators, and test them only in a dedicated test case as far as it's a corner case but it requires additional calls of `take` method, that makes the test a bit more complex.

```suggestion
      enum = Enumerator.product(1..2, ["A"], ["B"])
      enum.to_a.should == [[1, "A", "B"], [2, "A", "B"]]
      
      enum = Enumerator.product(2..3, ["A"], ["B"], ["C"])
      enum.to_a.should == [[2, "A", "B", "C"], [3, "A", "B", "C"]]
```

File: core/enumerator/product_spec.rb [hunk 3]
```diff
@@ -2,6 +2,29 @@
...
+    end
+
+    it "creates empty array with 0 arguments" do
+      enum = Enumerator.product
+      enum.to_a.should == [[]]
+    end
+
+    it "is an instance of Enumerator::Product" do
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:23]: **minor:**
```suggestion
    it "returns an instance of Enumerator::Product" do
```

File: core/enumerator/product_spec.rb [hunk 4]
```diff
@@ -2,6 +2,29 @@
...
+      enum = Enumerator.product
+      enum.class.should == Enumerator::Product
+    end
+
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:27]: **minor:**
regarding the test case below:

```ruby
    it "creates an infinite enumerator" do
      enum = Enumerator.product(1.., ["A", "B"])
      enum.take(5).should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"], [3, "A"]]
    end
```

I would reword the title and also check `#size` of the returned enumerator:

```ruby
    it "accepts infinite enumerators and returns infinite enumerator" do
      enum = Enumerator.product(1.., ["A", "B"])
      enum.take(5).should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"], [3, "A"]]
      enum.size.should == Float::INFINITY
    end
```

File: core/enumerator/product_spec.rb [hunk 5]
```diff
@@ -14,9 +37,33 @@
       elems.should == [[1, "X"], [1, "Y"], [2, "X"], [2, "Y"]]
     end
 
+    it "accepts infinite enumerator and a finite enumerable" do
+      enum = Enumerator.product(1.., "A".."C")
+      enum.size.should == Float::INFINITY
+      enum.take(4).should == [[1, "A"], [1, "B"], [1, "C"], [2, "A"]]
+    end
+
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:45]: **minor:** Wondering what is a purpose of these specs?

```ruby
it "accepts infinite enumerator and a finite enumerable" do
it "accepts infinite enumerator and uknown enumerator" do
it "accepts finite enumerator and unknown anumerator" do
```
Do you want to check that combinations of different enumerators are handled well? Or is it enough to check that each of them is handled properly in separate test cases?

File: core/enumerator/product_spec.rb [hunk 6]
```diff
@@ -14,9 +37,33 @@
...
+    it "reject keyword arguments" do
+      -> {
+        Enumerator.product(1..3, foo: 1, bar: 2)
+      }.should raise_error(ArgumentError)
+    end
+
+    it "can be iterated with #each_entry" do
+      enum = Enumerator.product(1..2, ["A", "B"])
+      enum.each_entry.to_a.should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"]]
+    end
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:67]: **minor:** Actually it's testing interface of the `Product` class. Moreover `Product` class inherits `#each_entry` from the `Enumerator` class and this method is defined in `Enumerable`... So there is nothing to check actually.

As far as I understand `#each_entry` method is called on arguments. So it's probably that it is the only requirement for the arguments. And they can be instances of any class, not only enumerable/enumerator, but any class that just implements the `#each_entry` method.

It's common to check such behaviour with `Object.new` e.g. we can check it in the following way:

```ruby
object = Object.new
def object.each_entry
  yield 1
  yield 2
end

Enumerator.product(object, ["A", "B"]).to_a.should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"]]
```

**Review by andrykonchin — COMMENTED** [review]
It looks good to me but there are some minor issues.

Also, please squash working commits before code review and after review but before merging a PR to have Git history clean and readable.

File: core/enumerator/product_spec.rb (hunk 5 above)
Comment (andrykonchin) [reply core/enumerator/product_spec.rb:45]: Ah, TBH I am not sure they are useful but it's OK to have them. MRI tests is a white box testing, so they might test corner cases specific to their implementation...

> What do you mean by to check that each of them is handled properly?
I meant to check that `Enumerator.product` handles these kinds of enumerators well.

File: core/enumerator/product_spec.rb [hunk 7]
```diff
@@ -2,9 +2,34 @@
...
+    it "creates a Cartesian product of enumerators" do
+      enum = Enumerator.product(1..2, ["A", "B"])
+      enum.to_a.should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"]]
+    end
+
+    it "accepts a list of enumerators of any length" do
+      enum = Enumerator.product(1..2, ["A"], ["B"])
+      enum.to_a.should == [[1, "A", "B"], [2, "A", "B"]]
+
+
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:14]: **code style:** extra empty line

File: core/enumerator/product_spec.rb [hunk 8]
```diff
@@ -14,9 +39,21 @@
...
-    it "accepts infinite enumerator and uknown enumerator" do
-      enum = Enumerator.product(1.., Enumerator.new { |y| y << "X" << "Y" })
-      enum.take(4).should == [[1, "X"], [1, "Y"], [2, "X"], [2, "Y"]]
+    it "reject keyword arguments" do
+      -> {
+        Enumerator.product(1..3, foo: 1, bar: 2)
+      }.should raise_error(ArgumentError)
+    end
+
+    it "can be iterated with #each_entry" do
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:48]: **minor:**
```suggestion
    it "requires argument to have only #each_entry method" do
```

**Review by andrykonchin — APPROVED** [review]
Looks good. There are some minor issues.

File: core/enumerator/product_spec.rb [hunk 9]
```diff
@@ -2,9 +2,34 @@
...
+
+
+      enum = Enumerator.product(2..3, ["A"], ["B"], ["C"])
+      enum.to_a.should == [[2, "A", "B", "C"], [3, "A", "B", "C"]]
+    end
+
+    it "creates an empty enumerator when called without arguments" do
+      enum = Enumerator.product
+      enum.to_a.should == [[]]
+    end
```
[the remark this answers is not in the document]
Comment (andrykonchin) [reply core/enumerator/product_spec.rb:22]: Looks good.

**Comment by andrykonchin** [comment]
I've pushed the last fixes into the Git branch
